autopush/web/validation.py

```diff
@@ -0,0 +1,166 @@
+    def validate(cls, schema):
+        """Validate a request schema in a separate thread before calling the
+        request handler
+
+        An alias `threaded_validate` should be used from this module.
+
+        Using `cyclone.web.asynchronous` is not needed as this function
+        will attach equivilant functionality to the method handler. Calling
+        `self.finish()` is needed on decorated handlers.
+
+        .. code-block::
+
+            class MyHandler(cyclone.web.RequestHandler):
+                @threaded_validate(MySchema())
+                def post(self):
+                    ...
+
+        """
+        return cls(schema)._decorator
+
+
+# Alias to the validation classmethod decorator
+threaded_validate = ThreadedValidate.validate
+
+
+class SimplepushSubscriptionSchema(Schema):
+    uaid = fields.UUID(required=True)
+    chid = fields.UUID(required=True)
+
+    @pre_load
+    def extract_subscription(self, d):
+        try:
+            result = self.context["settings"].parse_endpoint(
+                token=d["token"],
+                version=d["api_ver"],
+            )
+        except InvalidTokenException:
+            raise InvalidRequest("invalid token", errno=102)
+        return result
+
+    @validates_schema
+    def validate_uaid_chid(self, d):
+        try:
+            result = self.context["settings"].router.get_uaid(d["uaid"].hex)
+        except ItemNotFound:
+            raise InvalidRequest("UAID not found", status_code=410, errno=103)
+
+        if result.get("router_type") != "simplepush":
+            raise InvalidRequest("Wrong URL for user", errno=108)
+
+        # Propagate the looked up user data back out
+        d["user_data"] = result
+
+
+class SimplePushRequestSchema(Schema):
+    subscription = fields.Nested(SimplepushSubscriptionSchema,
+                                 load_from="token_info")
+    version = fields.Integer(missing=time.time)
+    data = fields.String(missing=None)
+
+    @validates('data')
+    def validate_data(self, value):
+        max_data = self.context["settings"].max_data
+        if value and len(value) > max_data:
+            raise InvalidRequest(
+                "Data payload must be smaller than {}".format(max_data),
+                errno=104,
+            )
+
+    @pre_load
+    def token_prep(self, d):
+        d["token_info"] = dict(
+            api_ver=d["path_kwargs"].get("api_ver"),
+            token=d["path_kwargs"].get("token"),
+        )
+        return d
+
+    @pre_load
+    def extract_fields(self, d):
+        body_string = d["body"]
+        version = data = None
+        if len(body_string) > 0:
+            body_args = urlparse.parse_qs(body_string, keep_blank_values=True)
+            version = body_args.get("version")
+            data = body_args.get("data")
+        else:
+            version = d["arguments"].get("version")
+            data = d["arguments"].get("data")
+        version = version[0] if version is not None else version
```

Comment:
It took a couple of reads to parse that this means `version = version[0] if version is not None else None`. How do you feel about writing this out as an `if` instead of a ternary?